RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1470 @@
+              Eigen::Map<const calo::multifit::ColumnVector<NSAMPLES>> noiseElectronicView{electronicNoiseTerms +
+                                                                                           gch * NSAMPLES};
+              Eigen::Map<const calo::multifit::ColMajorMatrix<NSAMPLES, NPULSES>> glbPulseMatrixMView{
+                  pulseMatricesM + gch * NSAMPLES * NPULSES};
+              Eigen::Map<const calo::multifit::ColMajorMatrix<NSAMPLES, NPULSES>> glbPulseMatrixPView{
+                  pulseMatricesP + gch * NSAMPLES * NPULSES};
+              Eigen::Map<const calo::multifit::ColMajorMatrix<NSAMPLES, NPULSES>> glbPulseMatrixView{
+                  pulseMatrices + gch * NSAMPLES * NPULSES};
+#ifdef HCAL_MAHI_GPUDEBUG
+              for (int i = 0; i < NSAMPLES; i++)
+                printf("inputValues(%d) = %f noiseTerms(%d) = %f\n", i, inputAmplitudesView(i), i, noiseTermsView(i));
+              for (int i = 0; i < NSAMPLES; i++) {
+                for (int j = 0; j < NPULSES; j++)
+                  printf("%f ", glbPulseMatrixView(i, j));
+                printf("\n");
+              }
+              printf("\n");
+              for (int i = 0; i < NSAMPLES; i++) {
+                for (int j = 0; j < NPULSES; j++)
+                  printf("%f ", glbPulseMatrixMView(i, j));
+                printf("\n");
+              }
+              printf("\n");
+              for (int i = 0; i < NSAMPLES; i++) {
+                for (int j = 0; j < NPULSES; j++)
+                  printf("%f ", glbPulseMatrixPView(i, j));
+                printf("\n");
+              }
+#endif
+
+              int npassive = 0;
+              float chi2 = 0, previous_chi2 = 0.f, chi2_2itersback = 0.f;
+              for (int iter = 1; iter < nMaxItersMin; iter++) {
+                //float covarianceMatrixStorage[MapSymM<float, NSAMPLES>::total];
+                // NOTE: only works when NSAMPLES == NPULSES
+                // if does not hold -> slightly rearrange shared mem to still reuse
+                // shared memory
+                float* covarianceMatrixStorage = shrMatrixLFnnlsStorage;
+                calo::multifit::MapSymM<float, NSAMPLES> covarianceMatrix{covarianceMatrixStorage};
+                CMS_UNROLL_LOOP
+                for (int counter = 0; counter < calo::multifit::MapSymM<float, NSAMPLES>::total; counter++)
+                  covarianceMatrixStorage[counter] = (noisecorr != 0.f) ? 0.f : averagePedestalWidth2;
+                CMS_UNROLL_LOOP
+                for (unsigned int counter = 0; counter < calo::multifit::MapSymM<float, NSAMPLES>::stride; counter++) {
+                  covarianceMatrix(counter, counter) += noiseTermsView.coeffRef(counter);
+                  if (counter != 0)
+                    covarianceMatrix(counter, counter - 1) +=
+                        noisecorr * noiseElectronicView.coeffRef(counter - 1) * noiseElectronicView.coeffRef(counter);
+                }
```

Comment:
I don't understand how this works :-(

`covarianceMatrix` is build in shared memory (from `shrMatrixLFnnlsStorage`), but different threads are writing to it and reading from it without any synchronisation ?